nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,588 @@
+    std::ostringstream buf;
+    Out buf_out(buf);
+    printValue(ctx, buf_out, *v.listElems()[n], "");
+    results.push_back(buf.str());
+  }
+  std::sort(results.begin(), results.end());
+  Out list_out(out, "[", "]", v.listSize());
+  for (auto const &v : results) {
+    list_out << v << Out::sep;
+  }
+}
+
+bool shouldSort(Context *ctx, Value &v) {
+  // Some lists should clearly be printed in sorted order, like
+  // environment.systemPackages.  Some clearly should not, like
+  // services.xserver.multitouch.buttonsMap.  As a conservative heuristic, sort
+  // lists of derivations.
+  return v.listSize() > 0 && ctx->state->isDerivation(*v.listElems()[0]);
+}
+
+void printList(Context *ctx, Out &out, Value &v) {
+  if (shouldSort(ctx, v)) {
+    printSortedList(ctx, out, v);
+  } else {
+    printUnsortedList(ctx, out, v);
+  }
+}
+
+void printAttrs(Context *ctx, Out &out, Value &v, std::string const &path) {
+  Out attrs_out(out, "{", "}", v.attrs->size());
+  for (const auto &a : v.attrs->lexicographicOrder()) {
+    std::string name = a->name;
+    attrs_out << name << " = ";
+    printValue(ctx, attrs_out, *a->value, appendPath(path, name));
+    attrs_out << ";" << Out::sep;
+  }
+}
+
+void multiLineStringEscape(Out &out, std::string const &s) {
+  int i;
+  for (i = 1; i < s.size(); i++) {
+    if (s[i - 1] == '$' && s[i] == '{') {
+      out << "''${";
+      i++;
+    } else if (s[i - 1] == '\'' && s[i] == '\'') {
+      out << "'''";
+      i++;
+    } else {
+      out << s[i - 1];
+    }
+  }
+  if (i == s.size()) {
+    out << s[i - 1];
+  }
+}
+
+void printMultiLineString(Out &out, Value const &v) {
+  std::string s = v.string.s;
+  Out str_out(out, "''", "''", Out::MULTI_LINE);
+  std::string::size_type begin = 0;
+  while (begin < s.size()) {
+    std::string::size_type end = s.find('\n', begin);
+    if (end == std::string::npos) {
+      multiLineStringEscape(str_out, s.substr(begin, s.size() - begin));
+      break;
+    }
+    multiLineStringEscape(str_out, s.substr(begin, end - begin));
+    str_out << Out::sep;
+    begin = end + 1;
+  }
+}
+
+void printValue(Context *ctx, Out &out, std::variant<Value, Error> maybe_value,
+                std::string const &path) {
+  try {
+    if (std::holds_alternative<Error>(maybe_value)) {
+      throw Error{std::get<Error>(maybe_value)};
+    }
+    Value v = evaluateValue(ctx, &std::get<Value>(maybe_value));
+    if (ctx->state->isDerivation(v)) {
+      describeDerivation(ctx, out, v);
+    } else if (v.isList()) {
+      printList(ctx, out, v);
+    } else if (v.type == tAttrs) {
+      printAttrs(ctx, out, v, path);
+    } else if (v.type == tString &&
+               std::string(v.string.s).find('\n') != std::string::npos) {
+      printMultiLineString(out, v);
+    } else {
+      ctx->state->forceValueDeep(v);
+      out << v;
+    }
+  } catch (Error &e) {
+    if (e.msg() == "The option `" + path + "' is used but not defined.") {
```

Comment:
Hm, matching on an error message is a bad idea. Maybe introduce a new exception type?

Reply:
I refined this from Error to ThrownError, but I don't know how to refine it further.  This is an error thrown from nix code (devel/nixpkgs/lib/modules.nix:329) not a place inside the nix interpreter where it would be straightforward to add another error type.

Reply:
I refined this from Error to ThrownError, but I don't know how to refine it further. This is an error thrown from nix code (devel/nixpkgs/lib/modules.nix:329) not a place inside the nix interpreter where it would be straightforward to add another error type.